src/cli/self_update/shell.rs

```diff
@@ -201,6 +206,46 @@ impl UnixShell for Zsh {
     }
 }
 
+struct Fish;
+
+impl UnixShell for Fish {
+    fn does_exist(&self) -> bool {
+        // fish has to either be the shell or be callable for fish setup.
+        matches!(process().var("SHELL"), Ok(sh) if sh.contains("fish"))
+            || matches!(utils::find_cmd(&["fish"]), Some(_))
+    }
+
+    // > "$XDG_CONFIG_HOME/fish/conf.d" (or "~/.config/fish/conf.d" if that variable is unset) for the user
+    // from <https://github.com/fish-shell/fish-shell/issues/3170#issuecomment-228311857>
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        let p0 = process().var("XDG_CONFIG_HOME").ok().map(|p| {
+            let mut path = PathBuf::from(p);
+            path.push("fish/config.d/rustup.fish");
+            path
+        });
+
+        let p1 = utils::home_dir().map(|mut path| {
+            path.push(".config/fish/config.d/rustup.fish");
+            path
+        });
+
+        p0.into_iter().chain(p1).collect()
+    }
+
+    fn update_rcs(&self) -> Vec<PathBuf> {
+        if let Ok(home) = process().var("XDG_CONFIG_HOME") {
+            let mut path = PathBuf::from(home);
+            path.push("fish/config.d/rustup.fish");
+            vec![path]
+        } else if let Some(mut path) = utils::home_dir() {
```

Comment:
Sorry for not getting to this earlier.

We generally try to write to all the places that a shell might read. The helper we wrote is idempotent to support this. What would happen if that was done here?

Reply:
Hm, to be honest I've lost all the context since I've opened this PR... Do you mean I should just call `rcfiles` here?

Reply:
approximately - look at zsh for the typical pattern